man/suauth.5
.\"     Title: suauth
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.70.1 <http://docbook.sf.net/>
.\"      Date: 07/30/2006
.\"    Manual: File Formats and Conversions
.\"    Source: File Formats and Conversions
.\"
.TH "SUAUTH" "5" "07/30/2006" "File Formats and Conversions" "File Formats and Conversions"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
suauth \- detailed su control file
.SH "SYNOPSIS"
.HP 12
\fB/etc/suauth\fR
.SH "DESCRIPTION"
.PP
The file
\fI/etc/suauth\fR
is referenced whenever the su command is called. It can change the behaviour of the su command, based upon:
.sp
.RS 3n
.nf
      1) the user su is targetting
    
.fi
.RE
.sp
.PP
2) the user executing the su command (or any groups he might be a member of)
.PP
The file is formatted like this, with lines starting with a # being treated as comment lines and ignored;
.sp
.RS 3n
.nf
      to\-id:from\-id:ACTION
    
.fi
.RE
.PP
Where to\-id is either the word
\fIALL\fR, a list of usernames delimited by "," or the words
\fIALL EXCEPT\fR
followed by a list of usernames delimited by ","
.PP
from\-id is formatted the same as to\-id except the extra word
\fIGROUP\fR
is recognised.
\fIALL EXCEPT GROUP\fR
is perfectly valid too. Following
\fIGROUP\fR
appears one or more group names, delimited by ",". It is not sufficient to have primary group id of the relevant group, an entry in
\fB/etc/group\fR(5)
is neccessary.
.PP
Action can be one only of the following currently supported options.
.TP 3n
\fIDENY\fR
The attempt to su is stopped before a password is even asked for.
.TP 3n
\fINOPASS\fR
The attempt to su is automatically successful; no password is asked for.
.TP 3n
\fIOWNPASS\fR
For the su command to be successful, the user must enter his or her own password. They are told this.
.PP
Note there are three separate fields delimited by a colon. No whitespace must surround this colon. Also note that the file is examined sequentially line by line, and the first applicable rule is used without examining the file further. This makes it possible for a system administrator to exercise as fine control as he or she wishes.
.SH "EXAMPLE"
.sp
.RS 3n
.nf
      # sample /etc/suauth file
      #
      # A couple of privileged usernames may
      # su to root with their own password.
      #
      root:chris,birddog:OWNPASS
      #
      # Anyone else may not su to root unless in
      # group wheel. This is how BSD does things.
      #
      root:ALL EXCEPT GROUP wheel:DENY
      #
      # Perhaps terry and birddog are accounts
      # owned by the same person.
      # Access can be arranged between them
      # with no password.
      #
      terry:birddog:NOPASS
      birddog:terry:NOPASS
      #
    
.fi
.RE
.sp
.SH "FILES"
.TP 3n
\fI/etc/suauth\fR
.SH "BUGS"
.PP
There could be plenty lurking. The file parser is particularly unforgiving about syntax errors, expecting no spurious whitespace (apart from beginning and end of lines), and a specific token delimiting different things.
.SH "DIAGNOSTICS"
.PP
An error parsing the file is reported using
\fBsyslogd\fR(8)
as level ERR on facility AUTH.
.SH "SEE ALSO"
.PP

\fBsu\fR(1).


NEWS
$Id: NEWS,v 1.492 2007/02/01 20:49:25 kloczek Exp $

shadow-4.0.18.1 -> shadow-4.0.18.2					13-10-2007

*** general:
- usermod: fixed handle -a option (by Benno Schulenberg
  <bensberg@justemail.net>),
- useradd: improved auditing support
  (https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=211659),
- groupadd, groupdel, groupmod, useradd, userdel, usermod: flush nscd cashes
  after close /etc/{group,passwd} files,
- su: If compiled without PAM support, enforce the limits from /etc/limits
  when one of the -, -l, or --login options is set, even if called by root.
*** documentation:
- updated translations: ja, nl, tl.
- groupadd.8, groupmod.8, login.1, useradd.8, userdel.8, usermod.8: grammar
  mistakes and other correctstions (by Schulenberg <bensberg@justemail.net>),

shadow-4.0.18 -> shadow-4.0.18.1					03-08-2006

*** general:
- groupmems: fixed compilation when PAM is disabled
  (by Johannes Winkelmann <jw@smts.ch>),
- fixed missing man pages in dist tar ball necessary on build when
  PAM is disabled.

shadow-4.0.17 -> shadow-4.0.18						01-08-2006

*** general:
- su: fixed set enviroment too early when using PAM, so move it to !USE_PAM
  (patch submitted by Mike Frysinger <vapier@gentoo.org>),
- groupadd, groupmod, useradd, usermod: fixed UID/GID overflow (fixed
  http://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=198920)
- passwd, useradd, usermod: fixed inactive/mindays/warndays/maxdays overflow
  (simillar to RH#198920),
- groupmems: rewrited for use PAM and getopt_long() and now it is enabled
  for build and install (patch by George Kraft <gk4@swbell.net>),
- S/Key: removed assign getpass() to libshadow_getpass() on autoconf level
  (patch by Ulrich Mueller <ulm@kph.uni-mainz.de>; http://bugs.gentoo.org/139966),
- usermod: back to previous -a option semantics and clarify -a behavior
  on documentation level (by Greg Schafer <gschafer@zip.com.au>),
- chsh, groupmod: rewrited for use getopt_long().
- updated translations: ca, cs, da, eu, fr, gl, hu, ko, pl, pt, ru, sv, tr, uk, vi.
*** documentation:
- fr and ru man pages are up to date,
- partially translated sv man pages set added
  (by Daniel Nylander <info@danielnylander.se>),
- pl chage(1), chsh(1), groupmod(8): translation finished.

shadow-4.0.16 -> shadow-4.0.17						10-07-2006

*** general:
- userdel, usermod: fixed segfault on remove home directory when it can't
  be removed; for example when it is /dev/null (fixed http://bugs.gentoo.org/139148),
- improved SELinux detection on autoconf level (based on patch by
  Dan Yefimov <dan@D00M.lightwave.net.ru>),
- removed using private implementation getpass() libc function
  (now getpass() is used also when S/KEY support is enabled),
- move nologin do $(sbindir),
- useradd: fixed mail spool file creation (bug cached by Frans Pop
  <elendil@planet.nl>;
  fixed http://bugs.debian.org/374705),
- updated translations: cs, da, de, ko, nb, nl, pt, ro, ru, sk, sv, vi,
- new translations: dz, km, ne.
*** documentation:
- ru man pages up to date,
- lastlog(8): updated pl translation,
- faillog(5): added missing information about fail_locktime element of
  faillog struct (by Thorsten Kukuk <kukuk@suse.de>),
- updated translations: eu, fr, pl.
- reverted using docbook.sourceforge.net in XSL url.

shadow-4.0.15 -> shadow-4.0.16						05-06-2006

*** general:
- userdel: better fix for old CERT VU#312962 (which was fixed in shadow 4.0.8):
  fixed forgoten checking of the return value from fchown() before
  proceeding with the fchmod() (based on Owl patch prepared by
  Rafal Wojtczuk <nergal@owl.openwall.com>),
- userdel: use login.defs::MAIL_DIR instead hardcoded /var/mail in created
  mailbox path (based on Owl fixes submited
  by Solar Designer <solar@openwall.com>),
- by default do not use libshadow_getpass() as getpass() replacemement.
  Use libshadow_getpass() only when S/KEY support is enabled.
  Current glibc getpass() handles correctly longer than 8 characters
  passwords and libshadow_getpass() is used only because libc getpass()
  do not handles password prompting with echo enabled,
- move login.defs::MD5_CRYPT_ENAB to non-PAM part,
- userdel: rewrited for use getopt_log(),
- install default/template configuration files:
-- if shadow is configured with use PAM install /etc/pam.d/* files,
-- if shadow do not uses PAM install /etc/{limits,login.acces} files,
-- install /etc/login.defs and /etc/default/useradd files,
- fixed handle relative symlinks too in lib/commonio.c
  (merge patch from Fedora),
- properly notify nscd to flush its cache
  (https://bugzilla.redhat.com/bugzilla/186803),
- useradd, usermod: fixes for verify return values mkdir() and chown()
  on copy files (merge 482_libmisc_copydir_check_return_values Debian
  patch),
- login, su (non-PAM variant): export MAIL only when MAIL_CHECK_ENAB
  is enabled (Mike Frysinger <vapier@gentoo.org>),
- pgck, grpck: warn when the members of a group differ in /etc/groups
  and /etc/gshadow (fixed http://bugs.debian.org/75181),
- su: fixed exit with a status 0 when the invoked command is terminated
  by a signal which was not catched
  (fixed by Eero Häkkinen <eero17@bigfoot.com>),
- login: cancel login timeout after authentication so that patient people
  timing out on network directory services can log in with local
  accounts (http://bugs.debian.org/107148),
- chgpasswd: fixes for build correctly with --disable-shadowgrp
  (patch by Johannes Winkelmann <jw@tks6.net>).
- updated translations: cs, da, es, eu, fi, fr, gl, hu, id, pt, ru, sk, sv, vi.
- new translations: hu.
*** documentation:
- new cs man pages: groupmems(8), groupmod(8), grpck(8), gshadow(5)
  (by Miroslav Kure <kurem@upcase.inf.upol.cz>),
- regenerate roff man pages using docbook-style-xsl-1.70.1,
- bunch of cleanups in chfn(1), faillog(8), gpasswd(1), groupadd(8),
  groupmems(8), limits(5), login(1), login.defs(5), newgrp(1), passwd(1),
  passwd(5) and su(1) (by Yuri Kozlov <kozlov.y@gmail.com>),
- update pl vipw(8) man page,
- added chgpasswd(8) ru man page,

src/suauth.c
#include <config.h>
#include <errno.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <sys/types.h>
#include "defines.h"
#include "prototypes.h"

#ifndef SUAUTHFILE
#define SUAUTHFILE "/etc/suauth"
#endif

#define	NOACTION	0
#define	NOPWORD		1
#define	DENY		-1
#define	OWNPWORD	2

/*
 * Global variables
 */
struct passwd pwent;

#ifdef SU_ACCESS

/* Really, I could do with a few const char's here defining all the 
 * strings output to the user or the syslog. -- chris
 */
static int applies (const char *, char *);

int check_su_auth (const char *, const char *);
int isgrp (const char *, const char *);

static int lines = 0;


int check_su_auth (const char *actual_id, const char *wanted_id)
{
	int posn, endline;
	const char field[] = ":";
	FILE *authfile_fd;
	char temp[1024];
	char *to_users;
	char *from_users;
	char *action;

	if (!(authfile_fd = fopen (SUAUTHFILE, "r"))) {
		/*
		 * If the file doesn't exist - default to the standard su
		 * behaviour (no access control).  If open fails for some
		 * other reason - maybe someone is trying to fool us with
		 * file descriptors limit etc., so deny access.  --marekm
		 */
		if (errno == ENOENT)
			return NOACTION;
		SYSLOG ((LOG_ERR,
			 "could not open/read config file '%s': %m\n",
			 SUAUTHFILE));
		return DENY;
	}

	while (fgets (temp, sizeof (temp), authfile_fd) != NULL) {
		lines++;

		if (temp[endline = strlen (temp) - 1] != '\n') {
			SYSLOG ((LOG_ERR,
				 "%s, line %d: line too long or missing newline",
				 SUAUTHFILE, lines));
			continue;
		}

		while (endline > 0 && (temp[endline - 1] == ' '
				       || temp[endline - 1] == '\t'
				       || temp[endline - 1] == '\n'))
			endline--;
		temp[endline] = '\0';

		posn = 0;
		while (temp[posn] == ' ' || temp[posn] == '\t')
			posn++;

		if (temp[posn] == '\n' || temp[posn] == '#'
		    || temp[posn] == '\0') {
			continue;
		}
		if (!(to_users = strtok (temp + posn, field))
		    || !(from_users = strtok ((char *) NULL, field))
		    || !(action = strtok ((char *) NULL, field))
		    || strtok ((char *) NULL, field)) {
			SYSLOG ((LOG_ERR,
				 "%s, line %d. Bad number of fields.\n",
				 SUAUTHFILE, lines));
			continue;
		}

		if (!applies (wanted_id, to_users))
			continue;
		if (!applies (actual_id, from_users))
			continue;
		if (!strcmp (action, "DENY")) {
			SYSLOG ((pwent.pw_uid ? LOG_NOTICE : LOG_WARN,
				 "DENIED su from `%s' to `%s' (%s)\n",
				 actual_id, wanted_id, SUAUTHFILE));
			fprintf (stderr,
				 _("Access to su to that account DENIED.\n"));
			fclose (authfile_fd);
			return DENY;
		} else if (!strcmp (action, "NOPASS")) {
			SYSLOG ((pwent.pw_uid ? LOG_INFO : LOG_NOTICE,
				 "NO password asked for su from `%s' to `%s' (%s)\n",
				 actual_id, wanted_id, SUAUTHFILE));
			fprintf (stderr,
				 _("Password authentication bypassed.\n"));
			fclose (authfile_fd);
			return NOPWORD;
		} else if (!strcmp (action, "OWNPASS")) {
			SYSLOG ((pwent.pw_uid ? LOG_INFO : LOG_NOTICE,
				 "su from `%s' to `%s': asking for user's own password (%s)\n",
				 actual_id, wanted_id, SUAUTHFILE));
			fprintf (stderr,
				 _
				 ("Please enter your OWN password as authentication.\n"));
			fclose (authfile_fd);
			return OWNPWORD;
		} else {
			SYSLOG ((LOG_ERR,
				 "%s, line %d: unrecognised action!\n",
				 SUAUTHFILE, lines));
		}
	}
	fclose (authfile_fd);
	return NOACTION;
}

static int applies (const char *single, char *list)
{
	const char split[] = ", ";
	char *tok;

	int state = 0;

	for (tok = strtok (list, split); tok != NULL;
	     tok = strtok (NULL, split)) {

		if (!strcmp (tok, "ALL")) {
			if (state != 0) {
				SYSLOG ((LOG_ERR,
					 "%s, line %d: ALL in bad place\n",
					 SUAUTHFILE, lines));
				return 0;
			}
			state = 1;
		} else if (!strcmp (tok, "EXCEPT")) {
			if (state != 1) {
				SYSLOG ((LOG_ERR,
					 "%s, line %d: EXCEPT in bas place\n",
					 SUAUTHFILE, lines));
				return 0;
			}
			state = 2;
		} else if (!strcmp (tok, "GROUP")) {
			if ((state != 0) && (state != 2)) {
				SYSLOG ((LOG_ERR,
					 "%s, line %d: GROUP in bad place\n",
					 SUAUTHFILE, lines));
				return 0;
			}
			state = (state == 0) ? 3 : 4;
		} else {
			switch (state) {
			case 0:	/* No control words yet */
				if (!strcmp (tok, single))
					return 1;
				break;
			case 1:	/* An all */
				SYSLOG ((LOG_ERR,
					 "%s, line %d: expect another token after ALL\n",
					 SUAUTHFILE, lines));
				return 0;
			case 2:	/* All except */
				if (!strcmp (tok, single))
					return 0;
				break;
			case 3:	/* Group */
				if (isgrp (single, tok))
					return 1;
				break;
			case 4:	/* All except group */
				if (isgrp (single, tok))
					return 0;
				/* FALL THRU */
			}
		}
	}
	if ((state != 0) && (state != 3))
		return 1;
	return 0;
}

int isgrp (const char *name, const char *group)
{
	struct group *grp;

	grp = getgrnam (group);

	if (!grp || !grp->gr_mem)
		return 0;

	return is_on_list (grp->gr_mem, name);
}
#endif				/* SU_ACCESS */
